components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/cache/AuthenticationContextLoader.java

```diff
@@ -0,0 +1,291 @@
+import org.wso2.carbon.identity.application.common.IdentityApplicationManagementException;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.common.model.ServiceProvider;
+import org.wso2.carbon.identity.application.mgt.ApplicationManagementServiceImpl;
+import org.wso2.carbon.idp.mgt.IdentityProviderManagementException;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is used to optimize the Authentication Context before storing it and again loaded it with objects.
+ */
+public class AuthenticationContextLoader {
+
+    private static final AuthenticationContextLoader instance = new AuthenticationContextLoader();
+    private static final Log log = LogFactory.getLog(AuthenticationContextLoader.class);
+
+    private AuthenticationContextLoader() { }
+
+    /**
+     * Singleton method.
+     * @return Authentication Context Loader
+     */
+    public static AuthenticationContextLoader getInstance() {
+
+        return instance;
+    }
+
+    /**
+     * This method is used to optimize the authentication context object.
+     * @param context Authentication context
+     * @throws AuthenticationContextLoaderException Authentication Context Loader Exception
+     */
+    public void optimizeAuthenticationContext(AuthenticationContext context)
+            throws AuthenticationContextLoaderException {
+
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Optimization process for the authentication context with context id : %s " +
+                    "has started", context.getContextIdentifier()));
+        }
+        optimizeExternalIdP(context);
+        optimizeAuthenticatorConfig(context);
+        optimizeApplicationConfig(context);
+    }
+
+    /**
+     * This method is used to load the authentication context from optimized authentication context's references.
+     * @param context Authentication context
+     * @throws AuthenticationContextLoaderException Authentication Context Loader Exception
+     */
+    public void loadAuthenticationContext(AuthenticationContext context) throws
+            AuthenticationContextLoaderException {
+
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Loading process for the authentication context with context id : %s " +
+                    "has started", context.getContextIdentifier()));
+        }
+        loadExternalIdP(context);
+        loadAuthenticatorConfig(context);
+        loadApplicationConfig(context);
+    }
+
+    private void optimizeExternalIdP(AuthenticationContext context) {
+
+        if (context.getExternalIdP() != null) {
+            context.setExternalIdPResourceId(context.getExternalIdP().getIdentityProvider().getResourceId());
```

Comment:
can `context.getExternalIdP().getIdentityProvider()` be null?

Reply:
We are constructing the external idp using the idp object, therefore it won't be null.